### PR TITLE
interfaces: maintain Plug and Slot connection details

### PR DESCRIPTION
**Pull request by zyga**
This branch is a continuation of the interface API changes where each Plug and Slot know about what they are connected to. In the past this data was only computed when Interfaces() was called and was actually not available at any other time. This branch changes that, making Interfaces() a trivial copy of what is already present and maintained by Connect() and Disconnect()

### CODE REVIEW

File: interfaces/repo.go [hunk 1]
```diff
@@ -400,6 +408,28 @@ func (r *Repository) disconnect(plug *Plug, slot *Slot) {
...
+		if plugRef.Snap == plug.Snap && plugRef.Name == plug.Name {
+			slot.Connections[i] = slot.Connections[len(slot.Connections)-1]
+			slot.Connections = slot.Connections[:len(slot.Connections)-1]
+			if len(slot.Connections) == 0 {
+				slot.Connections = nil
+			}
+			break
+		}
+	}
+	for i := 0; i < len(plug.Connections); i++ {
```
Comment (chipaca) [inline interfaces/repo.go]: why isn't this `for i, slotRef := range plug.Connections`?

File: interfaces/repo.go [hunk 2]
```diff
@@ -306,6 +306,14 @@ func (r *Repository) Connect(plugSnapName, plugName, slotSnapName, slotName stri
...
 	r.slotPlugs[slot][plug] = true
 	r.plugSlots[plug][slot] = true
+	slot.Connections = append(slot.Connections, PlugRef{
+		Snap: plug.Snap,
+		Name: plug.Name,
+	})
+	plug.Connections = append(plug.Connections, SlotRef{
+		Snap: slot.Snap,
+		Name: slot.Name,
+	})
```
Comment (niemeyer) [inline interfaces/repo.go]: Two lines only please.

File: interfaces/repo.go (hunk 1 above)
Comment (niemeyer) [inline interfaces/repo.go]: +1

File: interfaces/repo.go [hunk 3]
```diff
@@ -477,17 +507,11 @@ func (r *Repository) Interfaces() *Interfaces {
...
-			for slot := range r.plugSlots[plug] {
-				p.Connections = append(p.Connections, SlotRef{
-					Name: slot.Name,
-					Snap: slot.Snap,
-				})
+				Name:        plug.Name,
+				Snap:        plug.Snap,
+				Interface:   plug.Interface,
+				Label:       plug.Label,
+				Connections: plug.Connections[:],
```
Comment (niemeyer) [inline interfaces/repo.go]: The intention here seems to be copying data so the internal representation doesn't leak out, so this is a bug. The backing array is being shared. (not Python :-)

File: interfaces/repo.go [hunk 4]
```diff
@@ -497,17 +521,11 @@ func (r *Repository) Interfaces() *Interfaces {
...
-			for plug := range r.slotPlugs[slot] {
-				s.Connections = append(s.Connections, PlugRef{
-					Name: plug.Name,
-					Snap: plug.Snap,
-				})
+				Name:        slot.Name,
+				Snap:        slot.Snap,
+				Interface:   slot.Interface,
+				Label:       slot.Label,
+				Connections: slot.Connections[:],
```
Comment (niemeyer) [inline interfaces/repo.go]: Same.

File: interfaces/repo_test.go [hunk 5]
```diff
@@ -790,12 +790,77 @@ func (s *RepositorySuite) TestPlugConnectionsReturnsCorrectData(c *C) {
...
+				Interface: s.slot.Interface,
+				Label:     s.slot.Label,
+				Connections: []PlugRef{
+					{
+						Snap: s.plug.Snap,
+						Name: s.plug.Name,
+					},
+				},
+			},
+		},
```
Comment (niemeyer) [inline interfaces/repo_test.go]: Please use the compact representation I provided as an example in the follow up branches. There are about 10 lines too many here.

File: interfaces/repo_test.go [hunk 6]
```diff
@@ -790,12 +790,77 @@ func (s *RepositorySuite) TestPlugConnectionsReturnsCorrectData(c *C) {
...
+		Slots: []*Slot{
+			&Slot{
+				Snap:      s.slot.Snap,
+				Name:      s.slot.Name,
+				Interface: s.slot.Interface,
+				Label:     s.slot.Label,
+			},
+		},
+	})
```
Comment (niemeyer) [inline interfaces/repo_test.go]: Same.

**Comment by niemeyer** [comment]
LGTM with the comments addressed.

**Pull request merged**